lib/uri/common.rb

```diff
@@ -62,10 +64,36 @@ def make_components_hash(klass, array_hash)
 
   include REGEXP
 
-  @@schemes = {}
+  SCHEMES = {}
+  private_constant :SCHEMES
+
   # Returns a Hash of the defined schemes.
   def self.scheme_list
-    @@schemes
+    SCHEMES
+  end
+
+  # Registers a new scheme when adding custom URIs.
+  # Example:
+  #   module URI
+  #     class RSYNC < Generic
+  #       DEFAULT_PORT = 873
+  #     end
+  #     register_scheme('RSYNC', RSYNC)
+  #   end
+  def self.register_scheme(name, mod)
+    if SCHEMES.frozen?
```

Comment:
This branch is needed because anyone can load a new URI adapter (like the Rsync one) at any point.